src/components/sharing/util.js

```diff
@@ -0,0 +1,431 @@
+export const getUserMap = (responses, userInfos, resourceTotal) => {
+    const userMap = responses.reduce((acc, resp) => {
+        const type = getResponseType(resp);
+        const { permListFn, userIdFn, fmtSharingFn } = getSharingFns(type);
+        return resp[type].reduce(
+            (shares, sharing) => {
+                const permissions = permListFn(sharing);
+
+                return permissions.reduce(
+                    (userMap, permission) => {
+                        const userId = userIdFn(permission);
+                        const currentUser = {
+                            ...(userMap[userId] || userInfos[userId]),
+                        };
+                        const permissionValue = permission.permission;
+
+                        // Create an obj with the resource details and the user's
+                        // current permission
+                        const resourceWithPermission = {
+                            ...fmtSharingFn(sharing),
+                            permission: permissionValue,
+                        };
+
+                        const updatedUser = {
+                            ...currentUser,
+                            [type]: [
+                                ...(currentUser[type] || []),
+                                resourceWithPermission,
+                            ],
+                        };
+                        return { ...userMap, [userId]: updatedUser };
+                    },
+                    { ...shares }
+                );
+            },
+            { ...acc }
+        );
+    }, {});
+    return addDisplayPermissions(userMap, resourceTotal);
+};
+
+// Takes a user and examines its paths, apps, analyses, and tools keys to compare
+// against the resources list to find missing resources.  Adds the missing resources.
+export const addMissingResourcesToUser = (user, resources) => {
+    const currentResources = getUserResourceList(user);
+
+    return Object.keys(resources).reduce(
+        (acc, type) => {
+            const { idFn, fmtSharingFn } = getSharingFns(type);
+            return resources[type].reduce(
+                (resources, resource) => {
+                    const hasResource = currentResources.some(
+                        (perm) => idFn(perm) === idFn(resource)
+                    );
+                    if (!hasResource) {
+                        // Create new sharing object
+                        const updatedSharing = fmtSharingFn(resource);
+
+                        return {
+                            ...resources,
+                            [type]: [
+                                ...(resources[type] || []),
+                                updatedSharing,
+                            ],
+                        };
+                    } else {
+                        return { ...resources };
+                    }
+                },
+                { ...acc }
+            );
+        },
+        { ...user }
```

Comment:
This was a little tricky to follow, but I'm not sure what can be done about it: but it seems that the `user` is copied into the first reducer, which is then named `acc` and copied into the inner reducer, which now refers to it as `resources`.

In the end, since the inner reducer is always returning a copy of `user`/`resources`, then I don't think `user` and `acc` need to be copied initially.

Reply:
Reading this myself since seeing the comment intrigued me :sweat_smile:, I do think that the copies aren't needed too, and I also think that having the innermost reduce using the name `resources` is a likely source of confusion, since that's also an argument to the outermost function. I think there's some other stuff that might be more nicely done without the `...` operator too to make the whole thing shorter, which would I think help with reading it, despite the added code density

Perhaps something like:

```javascript
const addMissingResourcesToUser = (user, resources) => {
  const currentResources = getUserResourceList(user);

  return Object.keys(resources).reduce(
    (newUser, type) => {
      const { idFn, fmtSharingFn } = getSharingFns(type)
      return resources[type].reduce(
        (newResources, resource) => (
          currentResources.some((perm) => idFn(perm) === idFn(resource)) ?
          { ...newResources } :
          { ...newResources, [type]: (newResources[type] || []).concat(fmtSharingFn(resource)) }
        ),
        newUser
      )
      }
    },
    user
  )
}
```

(untested, but hopefully the ideas in terms of non-reuse of names + compactness come through anyway)

Reply:
Looks right to me 👍 
Although I'm fine with `...` spread/copy in place of `concat` as well:
```javascript
{ ...newResources, [type]: [...(newResources[type] || []), fmtSharingFn(resource)] }
```